cli/src/main/scala/scala/scalanative/cli/ScalaNativeP.scala

```diff
@@ -32,29 +39,79 @@ object ScalaNativeP extends CaseApp[POptions] {
       System.err.println(s"Ignoring non existing path: $path")
     }
 
+    if (options.fromPath) printFromFiles(classpath, args.all)
+    else printFromNames(classpath, args.all)
+  }
+
+  private def printFromNames(
+      classpath: List[Path],
+      args: Seq[String]
+  ): Unit = {
     Scope { implicit scope =>
       val classLoader =
         ClassLoader.fromDisk {
           Config.empty.withClassPath(classpath)
         }
 
       for {
-        fileName <- args.all
+        className <- args
       } {
-        classLoader.load(Global.Top(fileName)) match {
-          case Some(defns) =>
-            defns
-              .sortBy(_.name.mangle)
-              .foreach { d =>
-                println(d.show)
-                println()
-              }
-          case None => fail(s"Not found class/object with name `${fileName}`")
+        classLoader.load(Global.Top(className)) match {
+          case Some(defns) => printNIR(defns)
+          case None => fail(s"Not found class/object with name `${className}`")
         }
       }
     }
   }
 
+  private def printFromFiles(
+      classpath: List[Path],
+      args: Seq[String]
+  ): Unit = {
+
+    Scope { implicit scope =>
+      val cp = classpath.map(VirtualDirectory.real(_))
+      def isVirtualDirPathEqual(
+          virtualPath: Path,
+          regularPath: Path
+      ): Boolean = {
+        // Paths in jars are always using Unix path denotation
+        val relativeInJar = virtualPath.toString().stripPrefix("/")
+        relativeInJar == regularPath.toString()
+      }
+      def virtualDirHasPath(dir: VirtualDirectory, path: Path): Boolean = {
+        dir.files.exists(isVirtualDirPathEqual(_, path))
+      }
+      for {
+        fileName <- args
+        path = Paths.get(fileName).normalize()
+        content <- cp
+          .collectFirst {
+            case d if virtualDirHasPath(d, path) =>
+              // Paths in VirtualDirectories have a seperate, incomparable FileSystem
+              // Correct path has to be chosen from the read ones
+              val correctedPath = d.files.find { p =>
+                val relativeInJar = p.toString().stripPrefix("/")
```

Comment:
This lambda is the same as isVirtulDirPathEqual method, we can reuse it

Reply:
I don't like the fact that we do iterate files twice to find file, it might not be possible to fix it with standard collections methods as they would lead to iterationg over while claspath, so maybe let's change it. Maybe let's define a tail-recursive function that do something similar as the psedo-code below
Also potentially classpath can be replaced with Stream/LazyList
```
def findAndRead(classpath: List[VirtualDirectory]): Option[ByteBuffer] = {
  classpath match {
    case Nil => None
    case dir :: tail => 
     dir.files.find(matchesPath)
      .map(dir.read(_))
      .getOrElse(findAndRead(tail))
  }

}

```

Reply:
Great idea! I only changed the getOrElse to a simple if, as previously the compiler had trouble optimizing the recursion